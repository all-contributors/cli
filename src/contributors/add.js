const _ = require('lodash/fp')

function uniqueTypes(contribution) {
  return contribution.type || contribution
}

function formatContributions(options, existing, types) {
  if (options.url) {
    return (existing || []).concat(
      types.map(type => {
        return {type, url: options.url}
      }),
    )
  }
  return _.uniqBy(uniqueTypes, (existing || []).concat(types))
}

function updateContributor(options, contributor, contributions) {
  return _.assign(contributor, {
    contributions: formatContributions(
      options,
      contributor.contributions,
      contributions,
    ),
  })
}

function updateExistingContributor(options, username, contributions) {
  return options.contributors.map(contributor => {
    if (
      !contributor.login ||
      username.toLowerCase() !== contributor.login.toLowerCase()
    ) {
      return contributor
    }
    return updateContributor(options, contributor, contributions)
  })
}

function addNewContributor(options, username, contributions, infoFetcher) {
  return infoFetcher(username, options.repoType, options.repoHost).then(userData => {
    const contributor = _.assign(userData, {
      contributions: formatContributions(options, [], contributions),
    })
    return options.contributors.concat(contributor)
  })
}

module.exports = function addContributor(
  options,
  username,
  contributions,
  infoFetcher,
) {
  // case insensitive find
  const exists = _.find(contributor => {
    return (
      contributor.login &&
      contributor.login.toLowerCase() === username.toLowerCase()
    )
  }, options.contributors)

  if (exists) {
    return Promise.resolve(
      updateExistingContributor(options, username, contributions),
    )
  }
  return addNewContributor(options, username, contributions, infoFetcher)
}
