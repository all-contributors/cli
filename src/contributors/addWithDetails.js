const addContributor = require('./add')

// Adds a contributor without going to the network (you supply the additional fields)
module.exports = function addContributorWithDetails({
  options,
  login,
  contributions,
  name,
  avatar_url,
  profile,
}) {
  const infoFetcherNoNetwork = function() {
    return Promise.resolve({
      login,
      name,
      avatar_url,
      profile,
    })
  }
  return addContributor(options, login, contributions, infoFetcherNoNetwork)
}
