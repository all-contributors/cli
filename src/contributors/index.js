'use strict';

var _ = require('lodash/fp');
var util = require('../util');
var add = require('./add');
var github = require('./github');
var prompt = require('./prompt');

function isNewContributor(contributorList, username) {
  return !_.find({login: username}, contributorList);
}

module.exports = function addContributor(options, username, contributions) {
  const answersP = prompt(options, username, contributions);
  const contributorsP = answersP
    .then(answers => add(options, answers.username, answers.contributions, github));

  const writeContributorsP = contributorsP.then(
    contributors => util.configFile.writeContributors(options.config, contributors)
  );

  return Promise.all([answersP, contributorsP, writeContributorsP])
  .then(res => {
    const answers = res[0];
    const contributors = res[1];
    return {
      username: answers.username,
      contributions: answers.contributions,
      contributors: contributors,
      newContributor: isNewContributor(options.contributors, answers.username)
    };
  });
};
