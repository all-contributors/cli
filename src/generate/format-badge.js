'use strict';

var _ = require('lodash/fp');

var defaultTemplate = '[![All Contributors](https://img.shields.io/badge/all_contributors-<%= contributors.length %>-orange.svg?style=flat-square)](#contributors)';

module.exports = function formatBadge(options, contributors) {
  return _.template(options.badgeTemplate || defaultTemplate)({
    contributors: contributors
  });
};
