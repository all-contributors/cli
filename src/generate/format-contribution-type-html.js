const _ = require('lodash/fp')

const util = require('../util')

const linkTemplate = _.template(
  '<a href="<%= url %>" title="<%= description %>"><%= symbol %></a>',
)

function getType(options, contribution) {
  const types = util.contributionTypes(options)
  return types[contribution.type || contribution]
}

module.exports = function formatContribution(
  options,
  contributor,
  contribution,
) {
  const type = getType(options, contribution)

  if (!type) {
    throw new Error(
      `Unknown contribution type ${contribution} for contributor ${contributor.login ||
        contributor.name}`,
    )
  }

  const templateData = {
    symbol: type.symbol,
    description: type.description,
    contributor,
    options,
  }

  let url = getUrl(contribution, contributor)

  if (contribution.url) {
    url = contribution.url
  } else if (type.link) {
    url = _.template(type.link)(templateData)
  }

  return linkTemplate(_.assign({url}, templateData))
}

function getUrl(contribution, contributor) {
  if (contributor.login) {
    return `#${contribution}-${contributor.login}`
  } else {
    return `#${contribution}`
  }
}
