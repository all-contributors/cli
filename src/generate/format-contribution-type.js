const _ = require('lodash/fp')
const util = require('../util')

const linkTemplate = _.template(
  '[<%= symbol %>](<%= url %> "<%= description %>")',
)

function getType(options, contribution) {
  const types = util.contributionTypes(options)
  return types[contribution.type || contribution]
}

module.exports = function formatContribution(
  options,
  contributor,
  contribution,
) {
  const type = getType(options, contribution)

  if (!type) {
    throw new Error(
      `Unknown contribution type ${contribution} for contributor ${
        contributor.login
      }`,
    )
  }

  const templateData = {
    symbol: type.symbol,
    description: type.description,
    contributor,
    options,
  }

  let url = `#${contribution}-${contributor.login}`
  if (contribution.url) {
    url = contribution.url
  } else if (type.link) {
    url = _.template(type.link)(templateData)
  }

  return linkTemplate(_.assign({url}, templateData))
}
