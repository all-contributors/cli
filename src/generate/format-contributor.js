const _ = require('lodash/fp')
const formatContributionType = require('./format-contribution-type')

const avatarTemplate = _.template(
  '<img src="<%= contributor.avatar_url %>" width="<%= options.imageSize %>px;" height="<%= options.imageSize %>px;" alt="<%= name %>"/>',
)
const avatarBlockTemplate = _.template(
  '<a href="<%= contributor.profile %>"><%= avatar %><br /><sub><b><%= name %></b></sub></a>',
)
const avatarBlockTemplateNoProfile = _.template(
  '<%= avatar %><br /><sub><b><%= name %></b></sub>',
)
const contributorTemplate = _.template(
  '<%= avatarBlock %><br /><%= contributions %>',
)

const defaultImageSize = 100

function defaultTemplate(templateData) {
  const name = escapeName(templateData.contributor.name)
  const avatar = avatarTemplate(
    _.assign(templateData, {
      name,
    }),
  )
  const avatarBlockTemplateData = _.assign(
    {
      name,
      avatar,
    },
    templateData,
  )
  let avatarBlock = null

  if (templateData.contributor.profile) {
    avatarBlock = avatarBlockTemplate(avatarBlockTemplateData)
  } else {
    avatarBlock = avatarBlockTemplateNoProfile(avatarBlockTemplateData)
  }

  return contributorTemplate(_.assign({avatarBlock}, templateData))
}

function escapeName(name) {
  return name.replace(new RegExp('\\|', 'g'), '&#124;')
}

module.exports = function formatContributor(options, contributor) {
  const formatter = _.partial(formatContributionType, [options, contributor])
  const contributions = contributor.contributions.map(formatter).join(' ')
  const templateData = {
    contributions,
    contributor,
    options: _.assign({imageSize: defaultImageSize}, options),
  }
  const customTemplate =
    options.contributorTemplate && _.template(options.contributorTemplate)
  return (customTemplate || defaultTemplate)(templateData)
}
