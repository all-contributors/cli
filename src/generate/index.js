const _ = require('lodash/fp')
const formatBadge = require('./format-badge')
const formatContributor = require('./format-contributor')

function injectListBetweenTags(newContent) {
  return function(previousContent) {
    const tagToLookFor = `<!-- ALL-CONTRIBUTORS-LIST:`
    const closingTag = '-->'
    const startOfOpeningTagIndex = previousContent.indexOf(
      `${tagToLookFor}START`,
    )
    const endOfOpeningTagIndex = previousContent.indexOf(
      closingTag,
      startOfOpeningTagIndex,
    )
    const startOfClosingTagIndex = previousContent.indexOf(
      `${tagToLookFor}END`,
      endOfOpeningTagIndex,
    )
    if (
      startOfOpeningTagIndex === -1 ||
      endOfOpeningTagIndex === -1 ||
      startOfClosingTagIndex === -1
    ) {
      return previousContent
    }
    return [
      previousContent.slice(0, endOfOpeningTagIndex + closingTag.length),
      '\n<!-- prettier-ignore-start -->',
      '\n<!-- markdownlint-disable -->',
      newContent,
      '<!-- markdownlint-enable -->',
      '\n<!-- prettier-ignore-end -->',
      '\n',
      previousContent.slice(startOfClosingTagIndex),
    ].join('')
  }
}

function formatLine(contributors) {
  return `<td align="center">${contributors.join(
    '</td>\n    <td align="center">',
  )}</td>`
}

function generateContributorsList(options, contributors) {
  return _.flow(
    _.sortBy(function(contributor) {
      if (options.contributorsSortAlphabetically) {
        return contributor.name
      }
    }),
    _.map(function formatEveryContributor(contributor) {
      return formatContributor(options, contributor)
    }),
    _.chunk(options.contributorsPerLine),
    _.map(formatLine),
    _.join('\n  </tr>\n  <tr>\n    '),
    newContent => {
      return `\n<table>\n  <tr>\n    ${newContent}\n  </tr>\n</table>\n\n`
    },
  )(contributors)
}

function replaceBadge(newContent) {
  return function(previousContent) {
    const tagToLookFor = `<!-- ALL-CONTRIBUTORS-BADGE:`
    const closingTag = '-->'
    const startOfOpeningTagIndex = previousContent.indexOf(
      `${tagToLookFor}START`,
    )
    const endOfOpeningTagIndex = previousContent.indexOf(
      closingTag,
      startOfOpeningTagIndex,
    )
    const startOfClosingTagIndex = previousContent.indexOf(
      `${tagToLookFor}END`,
      endOfOpeningTagIndex,
    )
    if (
      startOfOpeningTagIndex === -1 ||
      endOfOpeningTagIndex === -1 ||
      startOfClosingTagIndex === -1
    ) {
      return previousContent
    }
    return [
      previousContent.slice(0, endOfOpeningTagIndex + closingTag.length),
      '\n',
      newContent,
      '\n',
      previousContent.slice(startOfClosingTagIndex),
    ].join('')
  }
}

module.exports = function generate(options, contributors, fileContent) {
  const contributorsList =
    contributors.length === 0
      ? '\n'
      : generateContributorsList(options, contributors)
  const badge = formatBadge(options, contributors)
  return _.flow(
    injectListBetweenTags(contributorsList),
    replaceBadge(badge),
  )(fileContent)
}
