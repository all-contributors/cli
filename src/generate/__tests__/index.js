import generate from '../'
import contributors from './fixtures/contributors.json'

function fixtures() {
  const options = {
    projectOwner: 'kentcdodds',
    projectName: 'all-contributors',
    imageSize: 100,
    contributorsPerLine: 5,
    contributors,
    contributorTemplate: '<%= contributor.name %> is awesome!',
  }

  const jfmengels = {
    login: 'jfmengels',
    name: 'Jeroen Engels',
    html_url: 'https://github.com/jfmengels',
    avatar_url: 'https://avatars.githubusercontent.com/u/3869412?v=3',
    contributions: ['doc'],
  }

  const content = [
    '# project',
    '',
    'Description',
    '',
    '## Contributors',
    'These people contributed to the project:',
    '<!-- ALL-CONTRIBUTORS-LIST:START -->FOO BAR BAZ<!-- ALL-CONTRIBUTORS-LIST:END -->',
    '',
    'Thanks a lot everyone!',
  ].join('\n')

  return {options, jfmengels, content}
}

test('replace the content between the ALL-CONTRIBUTORS-LIST tags by a table of contributors', () => {
  const {kentcdodds, bogas04} = contributors
  const {options, jfmengels, content} = fixtures()
  const contributorList = [kentcdodds, bogas04, jfmengels]
  const result = generate(options, contributorList, content)

  expect(result).toMatchSnapshot()
})

test('split contributors into multiples lines when there are too many', () => {
  const {kentcdodds} = contributors
  const {options, content} = fixtures()
  const contributorList = [
    kentcdodds,
    kentcdodds,
    kentcdodds,
    kentcdodds,
    kentcdodds,
    kentcdodds,
    kentcdodds,
  ]
  const result = generate(options, contributorList, content)

  expect(result).toMatchSnapshot()
})

test('sorts the list of contributors if contributorsSortAlphabetically=true', () => {
  const {kentcdodds, bogas04} = contributors
  const {options, jfmengels, content} = fixtures()

  const resultPreSorted = generate(
    options,
    [bogas04, jfmengels, kentcdodds],
    content,
  )

  options.contributorsSortAlphabetically = true
  const resultAutoSorted = generate(
    options,
    [jfmengels, kentcdodds, bogas04],
    content,
  )

  expect(resultPreSorted).toEqual(resultAutoSorted)
})

test('not inject anything if there is no tags to inject content in', () => {
  const {kentcdodds} = contributors
  const {options} = fixtures()
  const contributorList = [kentcdodds]
  const content = ['# project', '', 'Description', '', 'License: MIT'].join(
    '\n',
  )

  const result = generate(options, contributorList, content)
  expect(result).toBe(content)
})

test('not inject anything if start tag is malformed', () => {
  const {kentcdodds} = contributors
  const {options} = fixtures()
  const contributorList = [kentcdodds]
  const content = [
    '# project',
    '',
    'Description',
    '<!-- ALL-CONTRIBUTORS-LIST:SSSSSSSTART -->',
    '<!-- ALL-CONTRIBUTORS-LIST:END -->',
    '',
    'License: MIT',
  ].join('\n')

  const result = generate(options, contributorList, content)
  expect(result).toBe(content)
})

test('not inject anything if end tag is malformed', () => {
  const {kentcdodds} = contributors
  const {options} = fixtures()
  const contributorList = [kentcdodds]
  const content = [
    '# project',
    '',
    'Description',
    '<!-- ALL-CONTRIBUTORS-LIST:START -->',
    '<!-- ALL-CONTRIBUTORS-LIST:EEEEEEEND -->',
    '',
    'License: MIT',
  ].join('\n')

  const result = generate(options, contributorList, content)
  expect(result).toBe(content)
})

test('inject nothing if there are no contributors', () => {
  const {options, content} = fixtures()
  const contributorList = []
  const expected = [
    '# project',
    '',
    'Description',
    '',
    '## Contributors',
    'These people contributed to the project:',
    '<!-- ALL-CONTRIBUTORS-LIST:START -->',
    '<!-- prettier-ignore-start -->',
    '<!-- markdownlint-disable -->',
    '<!-- markdownlint-enable -->',
    '<!-- prettier-ignore-end -->',
    '<!-- ALL-CONTRIBUTORS-LIST:END -->',
    '',
    'Thanks a lot everyone!',
  ].join('\n')

  const result = generate(options, contributorList, content)

  expect(result).toBe(expected)
})

test('replace all-contributors badge if present', () => {
  const {kentcdodds} = contributors
  const {options} = fixtures()
  const contributorList = [kentcdodds]
  const content = [
    '# project',
    '',
    'Badges',
    [
      '[![version](https://img.shields.io/npm/v/all-contributors-cli.svg?style=flat-square)](http://npm.im/all-contributors-cli)',
      '<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->\n',
      '[![All Contributors](https://img.shields.io/badge/all_contributors-0-orange.svg?style=flat-square)](#contributors-)\n',
      '<!-- ALL-CONTRIBUTORS-BADGE:END -->\n',
      '[![version](https://img.shields.io/npm/v/all-contributors-cli.svg?style=flat-square)](http://npm.im/all-contributors-cli)',
    ].join(''),
    '',
    'License: MIT',
  ].join('\n')
  const expected = [
    '# project',
    '',
    'Badges',
    [
      '[![version](https://img.shields.io/npm/v/all-contributors-cli.svg?style=flat-square)](http://npm.im/all-contributors-cli)',
      '<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->\n',
      '[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors-)\n',
      '<!-- ALL-CONTRIBUTORS-BADGE:END -->\n',
      '[![version](https://img.shields.io/npm/v/all-contributors-cli.svg?style=flat-square)](http://npm.im/all-contributors-cli)',
    ].join(''),
    '',
    'License: MIT',
  ].join('\n')

  const result = generate(options, contributorList, content)

  expect(result).toBe(expected)
})
