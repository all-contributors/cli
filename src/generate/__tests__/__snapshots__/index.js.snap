// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace the content between the ALL-CONTRIBUTORS-LIST tags by a table of contributors 1`] = `
"# project

Description

## Contributors
These people contributed to the project:
<!-- ALL-CONTRIBUTORS-LIST:START -->
<!-- prettier-ignore -->
<table><tr><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Divjot Singh is awesome!</td><td align=\\"center\\">Jeroen Engels is awesome!</td></tr></table>
<!-- ALL-CONTRIBUTORS-LIST:END -->

Thanks a lot everyone!"
`;

exports[`split contributors into multiples lines when there are too many 1`] = `
"# project

Description

## Contributors
These people contributed to the project:
<!-- ALL-CONTRIBUTORS-LIST:START -->
<!-- prettier-ignore -->
<table><tr><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Kent C. Dodds is awesome!</td></tr><tr><td align=\\"center\\">Kent C. Dodds is awesome!</td><td align=\\"center\\">Kent C. Dodds is awesome!</td></tr></table>
<!-- ALL-CONTRIBUTORS-LIST:END -->

Thanks a lot everyone!"
`;
